// @aws-amplify/ui-react-storage/dist/esm/components/StorageBrowser/views/utils/tableResolvers/copyResolvers.mjs
import { noop, capitalize } from '@aws-amplify/ui';
import '../../../displayText/context.mjs';
import { isCopyViewDisplayTextKey } from '../../../displayText/utils.mjs';
import { STATUS_ICONS, STATUS_LABELS } from './constants.mjs';
import { getCopyOrDeleteCancelCellContent, getFileSize, getFileType, getCopyCellFolder } from './utils.mjs';

const COPY_TABLE_KEYS = [
    'name',
    'folder',
    'type',
    'size',
    'status',
    'cancel',
];
const getCopyCellKey = ({ key, item, }) => `${key}-${item.data.id}`;
const name = (data) => {
    const key = getCopyCellKey(data);
    const { item } = data;
    const text = item.data.fileKey;
    const icon = STATUS_ICONS[item.status];
    return { key, type: 'text', content: { icon, text } };
};
const folder = (data) => {
    const key = getCopyCellKey(data);
    const text = getCopyCellFolder(data.item);
    return { key, type: 'text', content: { text } };
};
const type = (data) => {
    const key = getCopyCellKey(data);
    const { fileKey } = data.item.data;
    const text = getFileType(fileKey);
    return { key, type: 'text', content: { text } };
};
const size = (data) => {
    const key = getCopyCellKey(data);
    const { size: value } = data.item.data;
    const displayValue = getFileSize(value);
    return { key, type: 'number', content: { value, displayValue } };
};
const status = (data) => {
    const key = getCopyCellKey(data);
    const { item: { status }, props: { displayText }, } = data;
    const statusLabelKey = STATUS_LABELS[status];
    const text = isCopyViewDisplayTextKey(statusLabelKey)
        ? displayText[statusLabelKey]
        : '';
    return { key, type: 'text', content: { text } };
};
const cancel = (data) => {
    const key = getCopyCellKey(data);
    const content = getCopyOrDeleteCancelCellContent(data);
    return { key, type: 'button', content };
};
const COPY_CELL_RESOLVERS = {
    name,
    folder,
    type,
    size,
    status,
    cancel,
    /**
     * @deprecated
     *
     * non-upload view tables do not include "progress" headers but include here to
     * keep TS happy as "progress" headers were included in display text interfaces
     * and cannot be removed from the tables without a breaking change
     */
    progress: noop,
};
const COPY_TABLE_RESOLVERS = {
    getCell: (data) => COPY_CELL_RESOLVERS[data.key](data),
    getHeader: ({ key, props: { displayText } }) => {
        const text = displayText[`tableColumn${capitalize(key)}Header`];
        if (key === 'cancel') {
            return { key, type: 'text', content: { text } };
        }
        return { key, type: 'sort', content: { label: text } };
    },
    getRowKey: ({ item }) => item.data.id,
};

export { COPY_TABLE_KEYS, COPY_TABLE_RESOLVERS };
